ch3/[429-641]CHN.txt

```diff
@@ -0,0 +1,423 @@
+【429】那些不是很有用的内容
+
+相反我们真正想要做的是
+
+当我们在这里匹配用户名时
+
+我们确实想要记住用户名是什么
+
+因为这是我们想要输出的内容
+
+
+
+【435】在正则表达式中，我们可以这样做
+
+使用匹配组
+
+匹配组是一种方式以表达
+
+我想要记住这个值并在之后使用
+
+在正则表达式中
+
+【440】任何括起来的表达式
+
+就是这样一个匹配组
+
+这样实际上我们这里已经有一个了
+
+这是第一组，现在我们在这儿再建第二组
+
+
+
+【445】注意这些括号并不影响匹配
+
+对吗？因为它们仅仅表达了
+
+将这个表达式作为一个单元
+
+但后面我们没做任何修改
+
+所以还是只匹配一次
+
+【450】然后匹配组有用的原因是
+
+
+
+你可以在 replacement 中回顾它们
+
+这样，在这里的 replacement 处
+
+我可以键入 \2
+
+【455】这是你指代匹配组编号的方法
+
+既然这样，我正在表达
+
+先匹配整行
+
+之后在 replacement 处放入你在第二个匹配组匹配到的值
+
+
+
+【460】好的，记住这是第一个捕获组
+
+这是第二个匹配组
+
+这给了我所有的用户名
+
+现在如果你回顾我们写的
+
+它很复杂，好的，它可能现在有意义了
+
+【465】当我们写过这个过程
+
+它为什么必须是它现在这个样子
+
+但是这似乎并不是很明显
+
+这些命令行是如何运行的
+
+这就是正则表达式调试器非常有用的地方了
+
+【470】我们这里有一个
+
+网上有很多，但是在这里我已经
+
+提前填好了这个我们刚刚用过的表达式
+
+注意到，它讲述了
+
+所有的匹配结果，
+
+【475】其实因为前面框框的尺寸现在这个窗口有一点小
+
+但是如果我……这里，这个注释说
+
+.* 匹配所有字符
+
+0 次到任意次
+
+后面是 Disconnected from
+
+【480】后面是一个匹配组
+
+之后滚轮你看到了所有的东西
+
+这是一点，但它还让你给测试串
+
+之后根据你给的每个测试串匹配这种模式
+
+
+
+【485】
+
+并且像这样给不同的匹配组着色
+
+这里，我们将用户作为一个匹配组，对吗？
+
+好的整个串匹配了对吗？
+
+整个串是蓝色的所以匹配完成
+
+【490】绿色部分是第一个匹配组
+
+红色是第二个匹配组
+
+这是第三个，
+
+因为 preauth 也被括号括起来了
+
+这会是一个调试你的正则表达式的有效方式
+
+【495】例如如果我放 Disconnected from ……
+
+我们这里新添一行
+
+如果我把 Disconnected from 作为用户名
+
+好吧现在这行已经有这个用户名了
+
+
+
+【500】我想想之前的
+
+你会注意到在这种模式下
+
+这不再是一个问题
+
+因为它已经匹配了用户名
+
+如果我们把这整行或这行变成用户名会发生什么
+
+【505】
+
+如现在展现的，让人迷惑，对吗？
+
+这就是正则表达式很难写对的地方了
+
+因为它现在在尽量的匹配
+
+它匹配了用户名出现的第一个位置
+
+【510】
+
+或者第一个 invalid 但在这种情况下换了第二个位置
+
+因为它是贪心的，我可以将它变为非贪心
+
+通过在这里加一个 ?
+
+所以如果你在 + 或者 * 后缀上 ?
+
+【515】它会变成非贪心匹配
+
+它就不会尽量匹配更多的字符
+
+之后你会看到
+
+串被正确解析
+
+因为 .* 匹配会在第一个 Disconnected from 处停止
+
+【520】
+
+这也正是被 SHH 发出的，
+
+实际出现在我们日志中的那一个
+
+正如你从目前为止的讲解所看出的
+
+
+
+【525】正则表达式会非常复杂
+
+并且你很可能会用到各种各样迷惑的修饰符
+
+
+
+在你的模式串中
+
+真正学会它们的唯一方式是从简单的表达式开始
+
+【530】之后把它们堆起来直到它匹配到你所需要的
+
+
+
+通常你仅仅是在做一些一次性的工作
+
+像这里我们破解用户名时
+
+你不需要去考虑所有特殊情况，对吗？
+
+【535】
+
+你不需要去想有人用 SSH 的用户名成功地
+
+匹配了你的登录名格式，这可能并不是问题
+
+
+
+因为你只是找出用户名
+
+【540】
+
+然而正则表达式确实很强大
+
+而且你要格外仔细当你做一些真正重要的事
+
+
+
+你有问题吗？
```

Comment:
🌚

Reply:
Yeah if you put the question mark in a ??? not only match a single line

I guess ??? you have your log like many lines it will just like find the 1st match and then stop looking

What is the effect adding question mark suffix you have user that had two copy of that log as username 

So this look like very much a security application and it is really...